File: problems/scaler/1-D_Array/decreasing_dishes.js
/**
 * Decreasing dishes
 *
 * Given an array of N positive integers representing the weights of ingredients in a dish.
 * Find the maximum possible sum of a subarray with decreasing weights.
 *
 * Problem Constraints
 * 1 <= N <= 105
 * 0 <= A[i] <= 105
 *
 * Sum of all elements of A <= 109
 */
function soln(A) {
  let maxSoFar = A[0];
  let tempSum = A[0];
  for (let i = 1; i < A.length; i++) {
    if (A[i] >= A[i - 1]) {
      tempSum = A[i];
    } else {
      tempSum += A[i];
    }
    maxSoFar = Math.max(maxSoFar, tempSum);
  }

  return maxSoFar;
}

File: src/main/java/com/dsa/problems/scaler/One_D_Array/decreasing_dishes.java
package com.dsa.problems.scaler.One_D_Array;

import java.util.ArrayList;
import java.util.Arrays;

public class decreasing_dishes {
  /**
   * Decreasing dishes
   *
   * Given an array of N positive integers representing the weights of ingredients in a dish.
   * Find the maximum possible sum of a subarray with decreasing weights.
   *
   * Problem Constraints
   * 1 <= N <= 105
   * 0 <= A[i] <= 105
   *
   * Sum of all elements of A <= 109
   */
  private static int decreasingDishes(ArrayList<Integer> A) {
    int sum = A.get(0);
    int tempSum = A.get(0);
    for(int i = 1; i < A.size(); i++) {
      if(A.get(i) < A.get(i - 1)) {
        tempSum += A.get(i);
      } else {
        tempSum = A.get(i);
      }
      sum = Math.max(tempSum, sum);
    }

    return sum;
  }

  public static void main(String[] args) {
    System.out.println(decreasingDishes(new ArrayList<>(Arrays.asList(3,2,1,4))));
  }
}
